### ISSUE TITLE
The calculation of width parameter?

### ISSUE THREAD

**Issue by bingothreed** [PetraVidnerova/rbf_for_tf2#2]
Hi, I am confuse about this implementation:

```python
def call(self, x):

        C = tf.expand_dims(self.centers, -1)  # inserts a dimension of 1
        H = tf.transpose(C-tf.transpose(x))  # matrix of differences
        return tf.exp(-self.betas * tf.math.reduce_sum(H**2, axis=1))
```

I learned RBF neural network from this [video](https://www.youtube.com/watch?v=OUtTI99uRf4)
There is a width parameter:
![image](https://user-images.githubusercontent.com/15133787/101628879-75ef8b00-3a5b-11eb-8de5-e9cd6b929d48.png)
and then 
![image](https://user-images.githubusercontent.com/15133787/101628959-90c1ff80-3a5b-11eb-8496-7f0f1fe2a841.png)

I haven't seen the use of width parameter in your code. So, I modefied the code like this
```python
def call(self, x):
        C = K.expand_dims(self.centers)
        XC = K.transpose(K.transpose(x)-C)     
        D = K.expand_dims(K.sqrt(K.mean(XC**2,axis=0)),0) 
        H = XC/D
        return K.exp(-self.betas * K.sum(H**2, axis=1))
```
The full code with experiments is [here](https://github.com/1165048017/BlogLearning/tree/master/RBF/RBFClassify).

My code overfit easily, but not with your code.  Am I wrong?

Thank you for solving my problem~~~ ^_^

**Comment by PetraVidnerova** [comment]
Hi, I just use beta insted of width 
ie. instead h = exp( - (x-c)^2 / r) , I use exp( - beta * (x-c)^2)
so it should be beta = 1/r

**Comment by PetraVidnerova** [comment]
I see now that you modified the code, I will have a close look on that later. 
Overfitting may be connected to too small width, with narraw gaussians you overfitt more easily than with wider gaussians.

**Comment by bingothreed** [comment]
> Hi, I just use beta insted of width
> ie. instead h = exp( - (x-c)^2 / r) , I use exp( - beta * (x-c)^2)
> so it should be beta = 1/r

@PetraVidnerova  Thanks for your reply~~~ You mean that the width is learned directly!
But look into the theory, the width has its own equation. There are two ways as implemeted in this [code](https://pythonmachinelearning.pro/using-neural-networks-for-regression-radial-basis-function-networks/) using numpy. 
So, I wonder that if it is possible to add these two ways in keras implementation or it is unnecessary to add them because of the learning skill by the neural network.

**Comment by PetraVidnerova** [comment]
Both widths and centers are tuned during learning in my implementation, since here the whole network is trained using back propagation process (there are many ways of RBF learning).  
In the simple example test.py, value 2.0 is provided, that means they are all initialized uniformly to this value.
But you can write you ouwn initializer and initilize the widths as you wish. Just instead beta=2.0 in the constructor, you should use 
beta = MyBetaInitializer().

If you do not want them to change during learning, it has to be changed: instead of trainable=True use trainable = False, in build method of RBFLayer.

**Comment by bingothreed** [comment]
> Both widths and centers are tuned during learning in my implementation, since here the whole network is trained using back propagation process (there are many ways of RBF learning).
> In the simple example test.py, value 2.0 is provided, that means they are all initialized uniformly to this value.
> But you can write you ouwn initializer and initilize the widths as you wish. Just instead beta=2.0 in the constructor, you should use
> beta = MyBetaInitializer().
> 
> If you do not want them to change during learning, it has to be changed: instead of trainable=True use trainable = False, in build method of RBFLayer.

I reviewed the theory again. The width equation is derived from center parameter. But in reality,there should be a width coefficient `df`, like this
![image](https://user-images.githubusercontent.com/15133787/101715468-7b3deb80-3ad6-11eb-9575-c1d1ddeed2bc.png)
So the learning parameter in RBF layer maybe the center and width coefficient.
But the problem is that, in your code the forward inference is :
![image](https://user-images.githubusercontent.com/15133787/101716591-8c87f780-3ad8-11eb-891e-fe4c891f5979.png)
Refering to the theory, it should be:
![image](https://user-images.githubusercontent.com/15133787/101716580-8265f900-3ad8-11eb-9f9b-865b5d700f24.png)
In these two equations, they will have different derivative for center parameter:
![image](https://user-images.githubusercontent.com/15133787/101716736-dcff5500-3ad8-11eb-90b6-751f4648f978.png)
This difference will lead to different update for center parameter during learning~~~
My opinion is that the parameter settings in your code is still the same as before,  but the inference function should be changed:
![image](https://user-images.githubusercontent.com/15133787/101717220-d4f3e500-3ad9-11eb-8539-7356b6b4efa4.png)

**Comment by PetraVidnerova** [comment]
That is great comment! You are perfectly right. 

Basically, there are several approaches to RBF network training. 
The basic one is to 1. set centers, 2. set widths, 3. calculate output weights, one by one. 

The training in my code uses backpropagation and tunes all network params simultaneously. Back prop needs random initialization, however center are widths are not initialized randomly but to meaningful values. I use random or k-means init for centers, and for widhts just uniform values.

You approach to widths suggest, that they depend on the values of centers. So the change of centers during learning should also influence the values of widths. If centers were constant, we can hide the term (sqr(1/n sum ...))  in beta, but since we are changing them, we should ad this to the computation of h2.

**Comment by bingothreed** [comment]
> That is great comment! You are perfectly right.
> 
> Basically, there are several approaches to RBF network training.
> The basic one is to 1. set centers, 2. set widths, 3. calculate output weights, one by one.
> 
> The training in my code uses backpropagation and tunes all network params simultaneously. Back prop needs random initialization, however center are widths are not initialized randomly but to meaningful values. I use random or k-means init for centers, and for widhts just uniform values.
> 
> You approach to widths suggest, that they depend on the values of centers. So the change of centers during learning should also influence the values of widths. If centers were constant, we can hide the term (sqr(1/n sum ...)) in beta, but since we are changing them, we should ad this to the computation of h2.

Yes，the K-means is only used to initial the center parameter but changed while training.
And then, if I specify the term (sqr(1/n sum ...)) explicitly, and modefied the code like that
![image](https://user-images.githubusercontent.com/15133787/101732045-8b19f780-3af7-11eb-8e32-d744e3df418a.png)
I get strange result:
![image](https://user-images.githubusercontent.com/15133787/101732405-23b07780-3af8-11eb-8f32-5803fe760e08.png)

But your code is right:
![image](https://user-images.githubusercontent.com/15133787/101732362-0d0a2080-3af8-11eb-9fd1-0328d29297e1.png)

I get smaller training loss than your implementation(1.5925e-04 vs 6.8916e-04) ,but get worse prediction results~~~~ T_T
It is really strange. Do I have a wrong modefication or misunderstand the theory? 
The attached code is my experiment with your data.  I would appreciate it if you have time to help me with that? Thank you very much.

[rbf_for_tf2.zip](https://github.com/PetraVidnerova/rbf_for_tf2/files/5670821/rbf_for_tf2.zip)

**Comment by PetraVidnerova** [comment]
Could you try smaller number of hidden units?

**Comment by bingothreed** [comment]
> Could you try smaller number of hidden units?

Firstly, I thought it got overfit. So, I changed the hidder units from 10 to 5. But the loss was still small and the result was still bad.
![image](https://user-images.githubusercontent.com/15133787/101749031-39c63400-3b08-11eb-8099-501d14e349a9.png)

Then I find it doesn't look like that the model gets overfit, because the test data is the same as train data in your code.
I just wonder why the modified code get smaller loss than yours, but get even worse predict result. In my mind, the smaller 'MSE' loss means the model fits the train data better, but the experiment seems not support it.

I checked that problem by adding the `evaluation` function:
![image](https://user-images.githubusercontent.com/15133787/101750331-d1785200-3b09-11eb-8d10-8e36eed468a2.png)

the last epoch and evaluate is shown like:
your code
```
Epoch 2000/2000
50/50 [==============================] - 0s 40us/sample - loss: 5.0368e-04 - mean_squared_error: 5.0368e-04
50/50 [==============================] - 0s 339us/sample - loss: 5.0352e-04 - mean_squared_error: 5.0352e-04
```
my code
```
Epoch 2000/2000
50/50 [==============================] - 0s 40us/sample - loss: 3.3930e-04 - mean_squared_error: 3.3930e-04
50/50 [==============================] - 0s 379us/sample - loss: 0.1876 - mean_squared_error: 0.1876
```
What a strange result of my code~~ T_T

**Comment by PetraVidnerova** [comment]
MSE is just square of difference from training data, so in this simple example it is really strange that you get smaller MSE. 
It looks something is going wrong. I will have to take a closer look on that. Thanks for you effort.

**Comment by PetraVidnerova** [comment]
And we should put some points apart to test set , since in this simple example only trainset is used and it is not correct approach.

**Comment by bingothreed** [comment]
In my first [code](https://github.com/1165048017/BlogLearning/blob/master/RBF/RBFClassify/GaussRBF.ipynb), I made two simple dataset; 
The training set is:
```
def test_data2(sample_number = 1000):
    all_data = np.random.rand(sample_number*2,2)
    data1 = all_data[all_data[...,0]>all_data[...,1]]
    data2 = all_data[all_data[...,0]<=all_data[...,1]]
    y1 = np.zeros((data1.shape[0],1))#class0
    y2 = np.ones((data2.shape[0],1))#class 1
    
    train_data = np.vstack((data1,data2))
    train_label = np.vstack((y1,y2))
    
    shuffle_idx = np.arange(sample_number*2)
    np.random.shuffle(shuffle_idx)
    
    train_data = train_data[shuffle_idx]
    train_label = train_label[shuffle_idx]
    return train_data,train_label
```
![image](https://user-images.githubusercontent.com/15133787/101755626-10110b00-3b10-11eb-8ea8-d554827a6876.png)


the test set is:
```
## create dataset
def test_data1(sample_number = 1000):
    mean0=[2,7]
    cov=np.mat([[1,0],[0,2]])
    data1=np.random.multivariate_normal(mean0,cov,sample_number)

    mean1=[8,3]
    cov=np.mat([[1,0],[0,2]])
    data2=np.random.multivariate_normal(mean1,cov,sample_number)
    
    y1 = np.zeros((sample_number,1)) # class 0
    y2 = np.ones((sample_number,1)) # class1
    
    train_data = np.vstack((data1,data2))
    train_label = np.vstack((y1,y2))
    
    shuffle_idx = np.arange(sample_number*2)
    np.random.shuffle(shuffle_idx)
    
    train_data = train_data[shuffle_idx]
    train_label = train_label[shuffle_idx]
    return train_data,train_label
```
![image](https://user-images.githubusercontent.com/15133787/101751849-bf97ae80-3b0b-11eb-8db6-8cd8b0360ca9.png)

The predicted result seems like no problem:
![image](https://user-images.githubusercontent.com/15133787/101752129-1604ed00-3b0c-11eb-9c31-0d9f07e56bc7.png)

But when I changed the test data to a line:
```
x1 = np.linspace(-2,12,1000)
x2 = np.linspace(-2,12,1000)
test_x = np.vstack((x1,x2)).T
```
![image](https://user-images.githubusercontent.com/15133787/101752277-48aee580-3b0c-11eb-9251-d595c4ba1134.png)

In this line dataset, your code get reasonable result:
![image](https://user-images.githubusercontent.com/15133787/101752392-6e3bef00-3b0c-11eb-987b-4a396b1b1285.png)
